### PR TITLE
Cleanup

### PR DESCRIPTION
**Pull request by Shamrock-Frost**
I changed how Ifs work a little and fixed the array literal parser (previously it couldn't parse `[1+1]`)

### CODE REVIEW

File: examples/merge_sort.tom [hunk 1]
```diff
@@ -66,13 +66,11 @@ let merge_sort(arr: [Int]):[Int] =
...
 let arrayofsize(n: Int):[Int] =
-    let arr = []
-    let i = 0
-    while n > i do
-        arr = arr ++ 0
-        i = i + 1
+    if n == 0 then
+        return []
+    else
+        return [0] ++ arrayofsize(n - 1)
```
Comment (tommy-mor) [inline examples/merge_sort.tom:72]: clever

File: examples/merge_sort.tom [hunk 2]
```diff
@@ -66,13 +66,11 @@ let merge_sort(arr: [Int]):[Int] =
...
     end
-    return arr
 end
 println("asdaaasdasd")
```
Comment (tommy-mor) [inline examples/merge_sort.tom:75]: kill this line

File: src/core/parser/TommyParser.kt [hunk 3]
```diff
@@ -96,10 +92,12 @@ class TommyParser : Grammar<List<AST>>() {
...
     private val varParser = idParser.map(::Var)
 
-    private val arrayLiteralParser = -LBRA and separatedTerms(parser(this::preLiteral), COMMA, acceptZero = true) and -RBRA map { LArray(it.toMutableList()) }
+    private val arrayLiteralParser = -LBRA and separatedTerms(parser(this::expr), COMMA, acceptZero = true) and -RBRA map { LArray(it.toMutableList()) }
 
     private val arrayGetParser :Parser<Expr> = varParser and -LBRA and parser(this::expr) and -RBRA map { (name, index) -> ArrayAccess(name, index) }
 
+    val literalParser = stringParser or numParser or boolParser or arrayLiteralParser
+
     //switch out preexper thing with parser(this::expr) later, if it works with preexpr
```
Comment (tommy-mor) [inline src/core/parser/TommyParser.kt:101]: you did this good job, delete comment

File: src/core/parser/TommyParser.kt [hunk 4]
```diff
@@ -240,10 +238,16 @@ class TommyParser : Grammar<List<AST>>() {
...
-    val ifParser = -IF and expr and -THEN and zeroOrMore(parser(this::astParser)) and
-             zeroOrMore(-ELSEIF and expr and -THEN and zeroOrMore(parser(this::astParser))) and optional(-ELSE and zeroOrMore(parser(this::astParser))) and -END map { (cond, body, elifs, elsebody) ->
-        If(cond, body, elsebody, elifs)
-    }
+    private val elseifParser = -ELSEIF and expr and -THEN and zeroOrMore(parser(this::astParser))
+    private val ifParser = -IF and expr and -THEN and zeroOrMore(parser(this::astParser)) and
+             zeroOrMore(elseifParser) and optional(-ELSE and zeroOrMore(parser(this::astParser))) and
+            -END map { (cond, body, elifs, elsebody) ->
+                val elseBranch : If? = elsebody?.let { Else(it) }
+                val elseIfs = elifs.foldRight(elseBranch) { (currCond, currBody), next ->
```
Comment (tommy-mor) [inline src/core/parser/TommyParser.kt:246]: clever, if parsing is much cleaner now

File: src/interpreter/treeWalker.kt [hunk 5]
```diff
@@ -228,29 +229,16 @@ fun runbody(body: List<AST>, environment: MutableMap<String, Tuple2<String?, Any
...
-                                            return runbody(b, environment)
-                                        }
-                                    } else typeError(a, "should be bool")
-                                }
-                            }
-                            if (curr.elseBranch != null) {
-                                return runbody(curr.elseBranch, environment)
-                            }
-                        } else typeError(curr.cond, "should be bool")
+                    is IfStep -> {
```
Comment (tommy-mor) [inline src/interpreter/treeWalker.kt:232]: much nicer now good work

**Review by tommy-mor — COMMENTED** [review]
NICE

File: examples/merge_sort.tom (hunk 1 above)
Comment (Shamrock-Frost) [reply examples/merge_sort.tom:72]: I mean I'm too used to haskell tbh

File: src/core/parser/TommyParser.kt (hunk 4 above)
Comment (Shamrock-Frost) [reply src/core/parser/TommyParser.kt:246]: yeah I didn't like  how we used List<Tuple2<Expr, Body>> for the elseifs and I figured we could do a pseudo-linked list for ifs instead

**Pull request merged**